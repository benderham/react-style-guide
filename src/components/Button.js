import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledButton = styled.button`
  border: solid 1px red;
`;

const Button = ({ label, onClick }) => (
  <StyledButton onClick={onClick}>{label}</StyledButton>
);

Button.propTypes = {
  /** Text shown on the button. */
  label: PropTypes.string.isRequired,
  /** Function passed to the button's 'onClick' event. */
  onClick: PropTypes.func.isRequired,
};

export default Button;
