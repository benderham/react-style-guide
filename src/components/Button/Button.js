import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledButton = styled.button`
  position: relative;
  padding: 0.6rem 1.8rem;
  font-size: 1.2rem;
  font-weight: 700;
  text-transform: lowercase;
  line-height: 1;
  color: ${props =>
    props.kind === 'secondary'
      ? props.theme.colors.medium
      : props.theme.colors.bright};
  background-color: transparent;
  border: solid 2px
    ${props =>
      props.kind === 'secondary'
        ? props.theme.colors.medium
        : props.theme.colors.bright};
  box-shadow: 0 0 1px rgba(0, 0, 0, 0);
  transform: perspective(1px) translateZ(0);
  transition: color 0.3s ease-out;
  cursor: pointer;

  &::before {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: ${props =>
      props.kind === 'secondary'
        ? props.theme.colors.medium
        : props.theme.colors.bright};
    border: solid 2px
      ${props =>
        props.kind === 'secondary'
          ? props.theme.colors.medium
          : props.theme.colors.bright};
    transform: scaleX(0);
    transform-origin: 0 50%;
    transition: transform 0.3s ease-out;
    content: '';
    z-index: -1;
  }

  &:hover,
  &:focus,
  &:active {
    color: ${props => props.theme.colors.white};

    &::before {
      transform: scaleX(1);
    }
  }

  &:disabled {
    cursor: default;
    opacity: 0.3;

    &:hover,
    &:focus,
    &:active {
      color: ${props =>
        props.kind === 'secondary'
          ? props.theme.colors.medium
          : props.theme.colors.bright};

      &::before {
        transform: scaleX(0);
      }
    }
  }
`;

const Button = ({ kind, disabled, label, onClick, type }) => (
  <StyledButton kind={kind} type={type} onClick={onClick} disabled={disabled}>
    {label}
  </StyledButton>
);

Button.defaultProps = {
  kind: 'primary',
  disabled: false,
  type: 'button',
};

Button.propTypes = {
  kind: PropTypes.oneOf(['primary', 'secondary']),
  disabled: PropTypes.bool,
  /** Text shown on the button. */
  label: PropTypes.string.isRequired,
  /** Function passed to the button's 'onClick' event. */
  onClick: PropTypes.func.isRequired,
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
};

export default Button;
