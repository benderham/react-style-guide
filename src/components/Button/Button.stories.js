import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';

import Button from './Button';

storiesOf('Button', module)
  .add(
    'Primary',
    withInfo(`
      Primary button style with text label.
    `)(() => <Button label="Button" onClick={action('button-click')} />),
  )
  .add(
    'Secondary',
    withInfo(`
      Secondary button style with text label.
    `)(() => (
      <Button
        label="Button"
        onClick={action('button-click')}
        kind="secondary"
      />
    )),
  )
  .add(
    'Disabled',
    withInfo(`
      Disabled button.
    `)(() => (
      <Button label="Button" onClick={action('button-click')} disabled />
    )),
  );
