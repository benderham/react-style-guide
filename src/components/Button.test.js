import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Button from './Button';

beforeAll(() => {
  jest.spyOn(global.console, 'error').mockImplementation(warning => {
    throw new Error(warning);
  });
});

afterAll(() => {
  jest.restoreAllMocks();
});

it('renders correctly without error', () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
  };
  const output = shallow(<Button {...props} />);
  expect(shallowToJson(output)).toMatchSnapshot();
});

it('should handle the click event', () => {
  const mockFn = jest.fn();
  const props = {
    label: 'Button',
    onClick: mockFn,
  };
  const output = shallow(<Button {...props} />);
  output.simulate('click');
  expect(mockFn).toHaveBeenCalled();
});
