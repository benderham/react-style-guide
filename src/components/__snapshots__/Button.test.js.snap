// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly without error 1`] = `
<styled.button
  onClick={[Function]}
>
  Button
</styled.button>
`;
